File: examples/browser/boards/portfolio-tracker/portfolio-tracker.js
/**
 * portfolio-tracker.js
 *
 * Runs the full T0-T4 lifecycle of the portfolio board demo.
 *
 * This is a BLACK-BOX client of board-live-cards CLI.
 * It only calls CLI commands and does NOT inspect board internals.
 *
 * Usage:
 *   node portfolio-tracker.js
 *   node portfolio-tracker.js --task-executor <path>
 */

import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const CARDS_TEMPLATE = path.join(__dirname, 'cards');
const REPO_ROOT = path.join(__dirname, '..', '..', '..', '..');
const CLI_WRAPPER = path.join(REPO_ROOT, 'board-live-cards-cli.js');
const CLI_TS = path.join(REPO_ROOT, 'src', 'cli', 'board-live-cards-cli.ts');
const CLI_JS = path.join(REPO_ROOT, 'dist', 'cli', 'board-live-cards-cli.js');
const TSX_CLI = path.join(REPO_ROOT, 'node_modules', 'tsx', 'dist', 'cli.mjs');

// Keep runtime artifacts out of the repository.
const RUNTIME_ROOT = fs.mkdtempSync(path.join(os.tmpdir(), 'portfolio-tracker-'));
const BOARD = path.join(RUNTIME_ROOT, 'board-runtime');
const CARDS = path.join(RUNTIME_ROOT, 'cards');
const TMP_FILE = path.join(BOARD, 'tmp_file1');

function parseArgs(argv) {
  let taskExecutor;
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === '--task-executor') {
      const value = argv[i + 1];
      if (!value || value.startsWith('--')) {
        console.error('[ERROR] Missing value for --task-executor');
        process.exit(1);
      }
      taskExecutor = value;
      i += 1;
      continue;
    }
    console.error(`[ERROR] Unknown argument: ${arg}`);
    process.exit(1);
  }
  return { taskExecutor };
}

const options = parseArgs(process.argv.slice(2));

console.log(`Runtime root: ${RUNTIME_ROOT}`);

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function cliCommand() {
  // Prefer node+tsx CLI on Windows to avoid flashing transient cmd windows.
  if (fs.existsSync(CLI_TS) && fs.existsSync(TSX_CLI)) {
    return { cmd: process.execPath, prefixArgs: [TSX_CLI, CLI_TS] };
  }
  if (fs.existsSync(CLI_WRAPPER)) {
    return { cmd: process.execPath, prefixArgs: [CLI_WRAPPER] };
  }
  if (fs.existsSync(CLI_JS)) {
    return { cmd: process.execPath, prefixArgs: [CLI_JS] };
  }
  return { cmd: process.execPath, prefixArgs: [CLI_WRAPPER] };
}

function runCli(args, capture = false) {
  const { cmd, prefixArgs } = cliCommand();
  const env = { ...process.env };
  // This demo needs real worker dispatch; suppressing spawn keeps source/inference tasks in running state.
  delete env.BOARD_LIVE_CARDS_NO_SPAWN;
  const result = spawnSync(cmd, [...prefixArgs, ...args], {
    stdio: capture ? 'pipe' : 'inherit',
    shell: false,
    windowsHide: true,
    env,
    encoding: capture ? 'utf-8' : undefined,
  });

  if (result.error) {
    console.error(`[ERROR] Failed to run CLI ${args[0]}: ${result.error.message}`);
    process.exit(1);
  }

  if (result.status !== 0) {
    if (capture && result.stdout) process.stdout.write(result.stdout);
    if (capture && result.stderr) process.stderr.write(result.stderr);
    console.error(`\n[ERROR] board-live-cards-cli ${args[0]} exited with status ${result.status}`);
    process.exit(1);
  }

  return capture ? result.stdout : undefined;
}

function cli(...args) {
  runCli(args, false);
}

/** Spawn CLI with JSON piped to stdin. */
function runCliWithInput(args, inputJson) {
  const { cmd, prefixArgs } = cliCommand();
  const env = { ...process.env };
  delete env.BOARD_LIVE_CARDS_NO_SPAWN;
  const result = spawnSync(cmd, [...prefixArgs, ...args], {
    input: inputJson,
    stdio: ['pipe', 'inherit', 'inherit'],
    shell: false,
    windowsHide: true,
    env,
    encoding: 'utf-8',
  });
  if (result.error) {
    console.error(`[ERROR] Failed to run CLI ${args[0]}: ${result.error.message}`);
    process.exit(1);
  }
  if (result.status !== 0) {
    console.error(`\n[ERROR] board-live-cards-cli ${args[0]} exited with status ${result.status}`);
    process.exit(1);
  }
}

/**
 * Write card to board's card store then upsert it.
 * Replaces old --card <file> flag (now removed from CLI).
 */
function upsertCardFromFile(boardDir, cardFilePath, restart = false) {
  const card = JSON.parse(fs.readFileSync(cardFilePath, 'utf-8'));
  const baseRef = `::fs-path::${boardDir}`;
  runCliWithInput(
    ['updates-in-card-store', '--base-ref', baseRef],
    JSON.stringify({ ops: [{ op: 'update', id: card.id, 'card-content': card }] }),
  );
  const args = ['upsert-card', '--base-ref', baseRef, '--card-id', card.id];
  if (restart) args.push('--restart');
  cli(...args);
}

/**
 * Expand a simple glob pattern (*.ext) and upsert each matching file.
 * Replaces old --card-glob <pattern> flag (now removed from CLI).
 */
function upsertCardsFromGlob(boardDir, globPattern) {
  const dir = path.dirname(globPattern);
  const basename = path.basename(globPattern);
  let files;
  if (basename.startsWith('*.')) {
    const ext = basename.slice(1);
    files = fs.readdirSync(dir).filter(f => f.endsWith(ext)).map(f => path.join(dir, f));
  } else if (basename === '*') {
    files = fs.readdirSync(dir).map(f => path.join(dir, f)).filter(f => fs.statSync(f).isFile());
  } else {
    files = [globPattern];
  }
  for (const file of files) {
    upsertCardFromFile(boardDir, file);
  }
}

/** Parse the JSON status response from the CLI. */
function parseStatusData(jsonText) {
  try {
    return JSON.parse(jsonText)?.data ?? null;
  } catch {
    return null;
  }
}

/**
 * Format the JSON status output as a human-readable table matching the
 * old text-mode status format:  "<status>  <name>  [<provides>]"
 */
function formatStatusTable(jsonText) {
  const statusData = parseStatusData(jsonText);
  if (!statusData?.cards) return jsonText;
  return statusData.cards
    .map(card => {
      const provides = card.provides_declared?.length
        ? `[${card.provides_declared.join(', ')}]`
        : '[]';
      return `${card.status}  ${card.name}  ${provides}`;
    })
    .join('\n') + '\n';
}

function statusText() {
  return runCli(['status', '--base-ref', `::fs-path::${BOARD}`], true) ?? '';
}

async function waitForAllCompleted(label, timeoutMs = 90000, pollMs = 500) {
  const start = Date.now();
  const includeInferenceCards = fs.existsSync(path.join(CARDS, 'portfolio-risk-assessment.json'))
    && fs.existsSync(path.join(CARDS, 'rebalancing-strategy.json'));

  while (Date.now() - start < timeoutMs) {
    const json = statusText();
    const statusData = parseStatusData(json);
    const cards = statusData?.cards ?? [];
    const isCompleted = (name) => cards.some(c => c.name === name && c.status === 'completed');

    const requiredCards = [
      isCompleted('portfolio-form'),
      isCompleted('price-fetch'),
      isCompleted('holdings-table'),
      isCompleted('portfolio-value'),
    ];
    if (includeInferenceCards) {
      requiredCards.push(
        isCompleted('portfolio-risk-assessment'),
        isCompleted('rebalancing-strategy'),
      );
    }

    if (requiredCards.every(Boolean)) {
      console.log(`${label}: all cards completed.`);
      return;
    }

    await sleep(pollMs);
  }

  console.error(`[ERROR] ${label}: timed out waiting for all cards to complete.`);
  console.error(formatStatusTable(statusText()));
  process.exit(1);
}

function writePrices(prices) {
  if (!fs.existsSync(BOARD)) {
    fs.mkdirSync(BOARD, { recursive: true });
  }
  fs.writeFileSync(TMP_FILE, JSON.stringify(prices), 'utf-8');
  console.log(`Wrote prices: ${JSON.stringify(prices)}`);
}

function setupRuntimeCards() {
  fs.rmSync(CARDS, { recursive: true, force: true });
  fs.mkdirSync(RUNTIME_ROOT, { recursive: true });
  fs.cpSync(CARDS_TEMPLATE, CARDS, { recursive: true });
  fs.copyFileSync(path.join(__dirname, 'fetch-prices.js'), path.join(RUNTIME_ROOT, 'fetch-prices.js'));
}

function printTaskExecutorLog() {
  console.log('\n=== Task Executor Log (board-dir) ===');
  const candidates = fs
    .readdirSync(BOARD, { withFileTypes: true })
    .filter(entry => entry.isFile() && entry.name.endsWith('.jsonl') && entry.name.includes('executor'))
    .map(entry => path.join(BOARD, entry.name));
  const taskExecutorLog = candidates.find(p => path.basename(p) === 'task-executor.jsonl') ?? candidates[0];
  if (!taskExecutorLog) {
    console.log(`No task executor log found in board-dir: ${BOARD}`);
    return;
  }

  console.log(`Log file: ${taskExecutorLog}`);
  const content = fs.readFileSync(taskExecutorLog, 'utf-8');
  process.stdout.write(content || '(empty)\n');
}

(async () => {
  setupRuntimeCards();

  console.log('\n=== T0: Init board ===');
  fs.rmSync(BOARD, { recursive: true, force: true });

  if (options.taskExecutor) {
    // Convert legacy "node /path/to/script" value to ExecutionRef JSON body
    let scriptPath = options.taskExecutor.trim();
    if (scriptPath.startsWith('node ')) scriptPath = scriptPath.slice(5).trim();
    const taskExecutorRef = { meta: 'task-executor', howToRun: 'local-node', whatToRun: `::fs-path::${scriptPath}` };
    runCliWithInput(
      ['init', '--base-ref', `::fs-path::${BOARD}`],
      JSON.stringify({ 'task-executor-ref': taskExecutorRef }),
    );
  } else {
    cli('init', '--base-ref', `::fs-path::${BOARD}`);
  }
  upsertCardsFromGlob(BOARD, path.join(CARDS, '*.json'));

  console.log('\n--- T0 Status (after upsert-card) ---');
  process.stdout.write(formatStatusTable(statusText()));

  console.log('\n=== T1: Writing market prices ===');
  writePrices({ AAPL: 198.50, MSFT: 425.30, GOOG: 178.90, AMZN: 192.40, TSLA: 168.75 });
  await waitForAllCompleted('T1');

  console.log('\n--- T1 Status ---');
  process.stdout.write(formatStatusTable(statusText()));

  console.log('\n=== T2: Adding GOOG (100 shares) ===');
  const portfolioFormPath = path.join(CARDS, 'portfolio-form.json');
  const portfolioFormV2 = {
    id: 'portfolio-form',
    meta: { title: 'Portfolio Holdings Form' },
    provides: [{ bindTo: 'holdings', ref: 'card_data.holdings' }],
    card_data: {
      holdings: [
        { symbol: 'AAPL', qty: 50 },
        { symbol: 'MSFT', qty: 30 },
        { symbol: 'GOOG', qty: 100 },
      ],
    },
    view: {
      elements: [
        { kind: 'table', label: 'Holdings', data: { bind: 'card_data.holdings', columns: ['symbol', 'qty'] } },
      ],
    },
  };
  fs.writeFileSync(portfolioFormPath, JSON.stringify(portfolioFormV2, null, 2));

  upsertCardFromFile(BOARD, portfolioFormPath, true);
  await sleep(500);
  writePrices({ AAPL: 198.50, MSFT: 425.30, GOOG: 178.90, AMZN: 192.40, TSLA: 168.75 });
  await waitForAllCompleted('T2');

  console.log('\n--- T2 Status ---');
  process.stdout.write(formatStatusTable(statusText()));

  console.log('\n=== T3: Force price refresh — AAPL now 205.00 ===');
  cli('retrigger', '--base-ref', `::fs-path::${BOARD}`, '--id', 'price-fetch');
  await sleep(500);
  writePrices({ AAPL: 205.00, MSFT: 425.30, GOOG: 178.90, AMZN: 192.40, TSLA: 168.75 });
  await waitForAllCompleted('T3');

  console.log('\n--- T3 Status ---');
  process.stdout.write(formatStatusTable(statusText()));

  console.log('\n=== T4: Rapid successive portfolio updates (3x queue stress) ===');
  fs.writeFileSync(TMP_FILE, '', 'utf-8');

  const portfolioFormV3 = {
    id: 'portfolio-form',
    meta: { title: 'Portfolio Holdings Form' },
    provides: [{ bindTo: 'holdings', ref: 'card_data.holdings' }],
    card_data: {
      holdings: [
        { symbol: 'AAPL', qty: 50 },
        { symbol: 'MSFT', qty: 30 },
        { symbol: 'GOOG', qty: 100 },
        { symbol: 'AMZN', qty: 40 },
      ],
    },
    view: {
      elements: [
        { kind: 'table', label: 'Holdings', data: { bind: 'card_data.holdings', columns: ['symbol', 'qty'] } },
      ],
    },
  };

  const portfolioFormV4 = {
    id: 'portfolio-form',
    meta: { title: 'Portfolio Holdings Form' },
    provides: [{ bindTo: 'holdings', ref: 'card_data.holdings' }],
    card_data: {
      holdings: [
        { symbol: 'AAPL', qty: 45 },
        { symbol: 'MSFT', qty: 30 },
        { symbol: 'GOOG', qty: 110 },
        { symbol: 'TSLA', qty: 60 },
      ],
    },
    view: {
      elements: [
        { kind: 'table', label: 'Holdings', data: { bind: 'card_data.holdings', columns: ['symbol', 'qty'] } },
      ],
    },
  };

  const portfolioFormV5 = {
    id: 'portfolio-form',
    meta: { title: 'Portfolio Holdings Form' },
    provides: [{ bindTo: 'holdings', ref: 'card_data.holdings' }],
    card_data: {
      holdings: [
        { symbol: 'AAPL', qty: 40 },
        { symbol: 'MSFT', qty: 35 },
        { symbol: 'GOOG', qty: 120 },
        { symbol: 'TSLA', qty: 70 },
      ],
    },
    view: {
      elements: [
        { kind: 'table', label: 'Holdings', data: { bind: 'card_data.holdings', columns: ['symbol', 'qty'] } },
      ],
    },
  };

  // First update starts a source fetch request.
  fs.writeFileSync(portfolioFormPath, JSON.stringify(portfolioFormV3, null, 2));
  upsertCardFromFile(BOARD, portfolioFormPath, true);

  // Immediate second update should queue a newer checksum while the first request is in-flight.
  fs.writeFileSync(portfolioFormPath, JSON.stringify(portfolioFormV4, null, 2));
  upsertCardFromFile(BOARD, portfolioFormPath, true);

  // Immediate third update should overwrite queued checksum (latest-state wins).
  fs.writeFileSync(portfolioFormPath, JSON.stringify(portfolioFormV5, null, 2));
  upsertCardFromFile(BOARD, portfolioFormPath, true);

  // 7) wait for first request, then 8) write response prices for update #1 tickers.
  // await readFetchRequest('T4 first fetch', ['AAPL', 'MSFT', 'GOOG', 'AMZN']);
  writePrices({ AAPL: 205.00, MSFT: 425.30, GOOG: 178.90, AMZN: 192.40 });
  await sleep(5000);

  // 9) wait for second request, then 10) write response prices for update #5 tickers.
  // await readFetchRequest('T4 second fetch', ['AAPL', 'MSFT', 'GOOG', 'TSLA']);
  writePrices({ AAPL: 206.00, MSFT: 426.00, GOOG: 179.50, TSLA: 169.20 });

  await waitForAllCompleted('T4');

  console.log('\n--- T4 Status ---');
  process.stdout.write(formatStatusTable(statusText()));

  console.log('\n=== T5: Final board status ===');
  process.stdout.write(formatStatusTable(statusText()));

  printTaskExecutorLog();

  console.log('\nPortfolio tracker completed successfully');
})();

File: examples/browser/boards/portfolio-tracker/portfolio-tracker.py
#!/usr/bin/env python3
"""portfolio-tracker.py — E2E orchestrator for the portfolio board demo.

Black-box CLI client. All board and card-store operations are performed by
shelling out to board-live-cards-cli and card-store-cli. No env vars. No
fallbacks or defaults.
"""

import copy
import json
import os
import shutil
import subprocess
import sys
import tempfile
import time

# ── Path resolution ────────────────────────────────────────────────────────────
_HERE = os.path.dirname(os.path.abspath(__file__))
_REPO_ROOT = os.path.normpath(os.path.join(_HERE, '..', '..', '..', '..'))

NODE = shutil.which('node')
if not NODE:
    print('[ERROR] node not found on PATH', file=sys.stderr)
    sys.exit(1)

BOARD_CLI       = os.path.join(_REPO_ROOT, 'board-live-cards-cli.js')
CARD_STORE_CLI  = os.path.join(_REPO_ROOT, 'card-store.js')
FETCH_PRICES_JS = os.path.join(_HERE, 'portfolio-tracker-fetch-prices.js')

# ── Runtime directories (under os.tmpdir()/experiment/) ───────────────────────
_TMP_BASE       = os.path.join(tempfile.gettempdir(), 'experiment')
CARDSTORE_DIR   = os.path.join(_TMP_BASE, 'cardstore')
BOARDRUNTIME_DIR = os.path.join(_TMP_BASE, 'boardruntime')
OUTPUTS_DIR     = os.path.join(_TMP_BASE, 'outputs')

CARDSTORE_REF    = f'::fs-path::{CARDSTORE_DIR}'
BOARDRUNTIME_REF = f'::fs-path::{BOARDRUNTIME_DIR}'
OUTPUTS_REF      = f'::fs-path::{OUTPUTS_DIR}'

# ── Inline card definitions ────────────────────────────────────────────────────
CARD_PORTFOLIO_FORM = {
    "id": "portfolio-form",
    "meta": {"title": "Portfolio Holdings Form"},
    "provides": [{"bindTo": "holdings", "ref": "card_data.holdings"}],
    "card_data": {"holdings": []},
    "view": {
        "elements": [
            {"kind": "table", "label": "Holdings",
             "data": {"bind": "card_data.holdings", "columns": ["symbol", "qty"]}}
        ]
    }
}

CARD_PRICE_FETCH = {
    "id": "price-fetch",
    "meta": {"title": "Fetch Market Prices"},
    "requires": ["holdings"],
    "provides": [{"bindTo": "prices", "ref": "fetched_sources.prices"}],
    "card_data": {},
    "source_defs": [{
        "kind": "mock-quotes",
        "bindTo": "prices",
        "outputFile": "prices.json",
        "projections": {"tickers": "requires.holdings.symbol"}
    }],
    "view": {
        "elements": [
            {"kind": "table", "label": "Market Prices",
             "data": {"bind": "fetched_sources.prices"}}
        ]
    }
}

CARD_HOLDINGS_TABLE = {
    "id": "holdings-table",
    "meta": {"title": "Holdings Table"},
    "requires": ["holdings", "prices"],
    "provides": [{"bindTo": "table", "ref": "computed_values.table"}],
    "card_data": {},
    "compute": [{
        "bindTo": "table",
        "expr": (
            '{ "rows": $map(requires.holdings, function($h) { '
            '{ "symbol": $h.symbol, "qty": $h.qty, '
            '"price": $lookup(requires.prices, $h.symbol), '
            '"value": $h.qty * $lookup(requires.prices, $h.symbol) } }) }'
        )
    }],
    "view": {
        "elements": [
            {"kind": "table", "label": "Portfolio Positions",
             "data": {"bind": "computed_values.table.rows",
                      "columns": ["symbol", "qty", "price", "value"]}}
        ]
    }
}

CARD_PORTFOLIO_VALUE = {
    "id": "portfolio-value",
    "meta": {"title": "Portfolio Total Value"},
    "requires": ["table"],
    "provides": [{"bindTo": "totalValue", "ref": "computed_values.totalValue"}],
    "card_data": {},
    "compute": [
        {"bindTo": "totalValue", "expr": "$sum(requires.table.rows.value)"}
    ],
    "view": {
        "elements": [
            {"kind": "metric", "label": "Total Portfolio Value",
             "data": {"bind": "computed_values.totalValue"}}
        ]
    }
}

# ── Helpers ────────────────────────────────────────────────────────────────────
def set_holdings(card_json: dict, holdings: dict) -> dict:
    card = copy.deepcopy(card_json)
    card["card_data"]["holdings"] = [
        {"symbol": symbol, "qty": qty}
        for symbol, qty in holdings.items()
    ]
    return card


def run_board(*args):
    subprocess.run([NODE, BOARD_CLI, *args], check=True, shell=False)


def run_board_with_input(*args, input_json: str):
    subprocess.run(
        [NODE, BOARD_CLI, *args],
        input=input_json, check=True, shell=False, text=True,
    )


def run_board_capture(*args) -> str:
    result = subprocess.run(
        [NODE, BOARD_CLI, *args],
        check=True, shell=False, capture_output=True, text=True,
    )
    return result.stdout


def run_card_store_set(card: dict):
    subprocess.run(
        [NODE, CARD_STORE_CLI, 'set', '--store-ref', CARDSTORE_REF],
        input=json.dumps(card), check=True, shell=False, text=True,
    )


def read_json(path: str):
    with open(path, encoding='utf-8') as f:
        return json.load(f)


def wait_for_completed(label: str, timeout_s: float = 90.0, poll_s: float = 0.5):
    required_names = {'portfolio-form', 'price-fetch', 'holdings-table', 'portfolio-value'}
    deadline = time.monotonic() + timeout_s
    while time.monotonic() < deadline:
        raw = run_board_capture('status', '--base-ref', BOARDRUNTIME_REF)
        data = json.loads(raw).get('data', {})
        cards = data.get('cards', [])
        completed = {c['name'] for c in cards if c.get('status') == 'completed'}
        if required_names.issubset(completed):
            print(f'[{label}] all cards completed.')
            return
        time.sleep(poll_s)
    # timed out — print status and exit
    raw = run_board_capture('status', '--base-ref', BOARDRUNTIME_REF)
    print(f'[ERROR] {label}: timed out waiting for all cards to complete.', file=sys.stderr)
    print(raw, file=sys.stderr)
    sys.exit(1)


# ── T0a — Create runtime directories ──────────────────────────────────────────
print('\n=== T0a: Create runtime directories ===')
if os.path.exists(_TMP_BASE):
    shutil.rmtree(_TMP_BASE)
for d in (CARDSTORE_DIR, BOARDRUNTIME_DIR, OUTPUTS_DIR):
    os.makedirs(d)
    print(f'  created: {d}')

# ── T0b — Init board ───────────────────────────────────────────────────────────
print('\n=== T0b: Init board ===')
_task_executor_body = json.dumps({
    'task-executor-ref': {
        'meta': 'task-executor',
        'howToRun': 'local-node',
        'whatToRun': f'::fs-path::{FETCH_PRICES_JS}',
    }
})
run_board_with_input(
    'init',
    '--base-ref', BOARDRUNTIME_REF,
    '--card-store-ref', CARDSTORE_REF,
    '--outputs-store-ref', OUTPUTS_REF,
    input_json=_task_executor_body,
)

# ── T0c — Set all cards into card store ────────────────────────────────────────
print('\n=== T0c: Set all cards into card store ===')
run_card_store_set(set_holdings(CARD_PORTFOLIO_FORM, {"AAPL": 50, "MSFT": 30}))
run_card_store_set(CARD_PRICE_FETCH)
run_card_store_set(CARD_HOLDINGS_TABLE)
run_card_store_set(CARD_PORTFOLIO_VALUE)

# ── T0d — Upsert cards to board ────────────────────────────────────────────────
print('\n=== T0d: Upsert cards to board ===')
for _card_id in ('portfolio-form', 'price-fetch', 'holdings-table', 'portfolio-value'):
    run_board('upsert-card', '--base-ref', BOARDRUNTIME_REF, '--card-id', _card_id)

# ── T1 — Wait for all cards completed ──────────────────────────────────────────
print('\n=== T1: Wait for all cards completed ===')
wait_for_completed('T1')

_prices_path = os.path.join(OUTPUTS_DIR, 'data-objects', 'prices.json')
_prices_t1 = read_json(_prices_path)
assert isinstance(_prices_t1, dict) and len(_prices_t1) > 0, \
    'T1: prices.json is empty or not an object'
assert set(_prices_t1.keys()) == {'AAPL', 'MSFT'}, \
    f'T1: expected keys {{AAPL, MSFT}}, got {set(_prices_t1.keys())}'
assert all(isinstance(v, (int, float)) for v in _prices_t1.values()), \
    'T1: all price values must be numbers'
print('[T1] assertion passed: prices.json has AAPL, MSFT with numeric values.')

# ── T2a — Update holdings (GOOG added) ────────────────────────────────────────
print('\n=== T2a: Update holdings (GOOG added) ===')
run_card_store_set(set_holdings(CARD_PORTFOLIO_FORM, {"AAPL": 50, "MSFT": 30, "GOOG": 100}))

# ── T2b — Upsert portfolio-form with --restart ─────────────────────────────────
print('\n=== T2b: Upsert portfolio-form --restart ===')
run_board('upsert-card', '--base-ref', BOARDRUNTIME_REF,
          '--card-id', 'portfolio-form', '--restart')

# ── T2c — Wait and assert ──────────────────────────────────────────────────────
print('\n=== T2c: Wait for all cards completed ===')
wait_for_completed('T2c')

_prices_t2c = read_json(_prices_path)
assert set(_prices_t2c.keys()) == {'AAPL', 'MSFT', 'GOOG'}, \
    f'T2c: expected keys {{AAPL, MSFT, GOOG}}, got {set(_prices_t2c.keys())}'

_ht_cv_path = os.path.join(OUTPUTS_DIR, 'cards', 'holdings-table', 'computed_values.json')
_ht_cv_t2c = read_json(_ht_cv_path)
assert len(_ht_cv_t2c['table']['rows']) == 3, \
    f'T2c: expected 3 rows in holdings-table, got {len(_ht_cv_t2c["table"]["rows"])}'
print('[T2c] assertions passed: 3 tickers in prices, 3 rows in holdings-table.')

# ── T3 — Retrigger price-fetch, wait ──────────────────────────────────────────
print('\n=== T3: Retrigger price-fetch ===')
run_board('retrigger', '--base-ref', BOARDRUNTIME_REF, '--id', 'price-fetch')
wait_for_completed('T3')

_prices_t3 = read_json(_prices_path)
assert set(_prices_t3.keys()) == {'AAPL', 'MSFT', 'GOOG'}, \
    f'T3: expected 3 tickers, got {set(_prices_t3.keys())}'
assert _prices_t3 != _prices_t2c, \
    'T3: prices must differ from T2c values after retrigger (random regeneration)'
print('[T3] assertions passed: 3 tickers, prices differ from T2c.')

# ── T4 — Rapid 5× portfolio-form updates (queue stress test) ──────────────────
print('\n=== T4: Rapid 5x portfolio-form updates ===')
for _holdings in [
    {"AAPL": 50, "MSFT": 30, "GOOG": 100, "AMZN": 40},              # V3
    {"AAPL": 45, "MSFT": 30, "GOOG": 110, "AMZN": 40, "TSLA": 60},  # V4
    {"AAPL": 45, "MSFT": 30, "GOOG": 110, "AMZN": 100},              # V4a
    {"AAPL": 45, "MSFT": 30, "GOOG": 110, "AMZN": 140, "TSLA": 60}, # V4b
    {"AAPL": 40, "MSFT": 35, "GOOG": 120, "TSLA": 70},               # V5
]:
    run_card_store_set(set_holdings(CARD_PORTFOLIO_FORM, _holdings))
    run_board('upsert-card', '--base-ref', BOARDRUNTIME_REF,
              '--card-id', 'portfolio-form', '--restart')

wait_for_completed('T4')

_prices_t4 = read_json(_prices_path)
assert set(_prices_t4.keys()) == {'AAPL', 'MSFT', 'GOOG', 'TSLA'}, \
    f'T4: expected keys {{AAPL, MSFT, GOOG, TSLA}}, got {set(_prices_t4.keys())}'
assert 'AMZN' not in _prices_t4, \
    'T4: AMZN must not be present (board must have settled on V5 holdings)'
print('[T4] assertions passed: V5 tickers only, AMZN absent.')

# ── T5 — Print final status and cross-check ────────────────────────────────────
print('\n=== T5: Print final status and cross-check ===')

# Step 1: Wait for all cards completed (stable state before any assertions)
wait_for_completed('T5')

# Step 2: Capture live CLI status
_cli_raw = run_board_capture('status', '--base-ref', BOARDRUNTIME_REF)
_cli_status = json.loads(_cli_raw)['data']

# Step 3: Read status.json from outputs store
_file_status = read_json(os.path.join(OUTPUTS_DIR, 'status.json'))

# Step 4: Cross-check CLI vs file status
assert json.dumps(_cli_status, sort_keys=True) == json.dumps(_file_status, sort_keys=True), \
    'T5: CLI status does not match status.json snapshot'
print('[T5] cross-check passed: CLI status matches status.json.')

# Step 5: Print holdings-table computed values
_ht_cv = read_json(_ht_cv_path)
print('\nFinal portfolio positions table:')
print(json.dumps(_ht_cv['table'], indent=2))

# Step 6: Totals cross-verify: holdings × prices == totalValue
V5_HOLDINGS = {"AAPL": 40, "MSFT": 35, "GOOG": 120, "TSLA": 70}
_prices_final = read_json(_prices_path)
_pv_cv = read_json(
    os.path.join(OUTPUTS_DIR, 'cards', 'portfolio-value', 'computed_values.json')
)
_total_value = _pv_cv['totalValue']
_expected = sum(qty * _prices_final[sym] for sym, qty in V5_HOLDINGS.items())
assert round(_expected, 2) == round(_total_value, 2), \
    f'T5: totals mismatch: expected={round(_expected, 2)}, got={round(_total_value, 2)}'
print(f'[T5] totals assertion passed: expected={round(_expected, 2)}, totalValue={round(_total_value, 2)}')

# Step 7: Print full CLI status
print('\nFinal board status:')
print(json.dumps(_cli_status, indent=2))

print('\n=== portfolio-tracker completed successfully ===')
